base/inc/AdePT/SparseVector.h

```diff
@@ -0,0 +1,746 @@
+      Mask_t mask = svect->mask_at(im).load();
+      for (auto i = 0; i < 32; ++i) {
+        if ((mask & (1 << (i & 31))) && pred_func(32 * im + i, svect)) {
+          int idest = dest_offset + (*num_selected)++;
+          new (output->data() + idest) Type((*svect)[32 * im + i]);
+          output->set_bit(idest);
+          mask &= ~(1 << (i & 31)); // remove selected element from mask
+          svect->mask_at(im).store(mask);
+        }
+      }
+    }
+    output->fNused += *num_selected;
+    output->fNshared += *num_selected;
+    svect->fNused -= *num_selected;
+  }
+#endif
+
+  template <typename Container>
+  static void select_used(const Vector_t *svect, Container *output, unsigned *num_selected)
+  {
+    Vector_t::select(svect, [] __device__(int, const Vector_t *) { return true; }, output, num_selected);
+  }
+
+  /** @brief Compacts used elements of this vect into a destination vect. */
+#ifdef COPCORE_CUDA_COMPILER
+  static void compact(Vector_t *svect, Vector_t *output, unsigned *num_selected)
+  {
+    // copy to destination vector
+    sa_detail::select_and_move_kernel_launcher<Type>
+        <<<1, 1>>>(svect, [] __device__(int, const Vector_t *) { return true; }, output, num_selected);
+    // update output vector
+    sa_detail::clear_kernel<Type><<<1, 1>>>(svect);
+  }
+#else
+  static void compact(Vector_t *svect, Vector_t *output, unsigned *num_selected)
+  {
+    // host version
+    // copy to destination vector
+    int mask_size         = svect->size() / 32;
+    const int dest_offset = output->size();
+    *num_selected         = 0;
+    for (auto im = 0; im < mask_size; ++im) {
+      Mask_t mask = svect->mask_at(im).load();
+      for (auto i = 0; i < 32; ++i) {
+        if (mask & (1 << (i & 31))) new (output->data() + dest_offset + (*num_selected)++) Type((*svect)[32 * im + i]);
+      }
+    }
+    output->fNused += *num_selected;
+    output->fNshared += *num_selected;
+    svect->clear();
+  }
+#endif
+
+  /** @brief Fills selection vector fSelected with indices of remaining used elements. */
+#ifdef COPCORE_CUDA_COMPILER
+  static void release_selected(Vector_t *svect, unsigned *selection, unsigned *nselected)
+  {
+    // we pass n_elements by pointer to avoid having to copy the value to the host
+    sa_detail::release_selected_kernel_launcher<Type><<<1, 1>>>(svect, selection, nselected);
+  }
+#else
+  static void release_selected(Vector_t *svect, unsigned *selection, unsigned *nselected)
+  {
+    // host version
+    for (int i = 0; i < *nselected; ++i)
+      svect->release(selection[i]);
+  }
+#endif
+};
+
+/** @brief Storage class for SparseVector */
+template <typename Type, unsigned N>
+struct SparseVectorStorage {
+  alignas(Type) char fData[N * sizeof(Type)];
+  unsigned int fMasks[N / 32];
+};
+
+/// Aligned storage may be needed also in case of SparseVector of size 0, if allocating in a buffer
+template <typename Type>
+struct alignas(Type) SparseVectorStorage<Type, 0> {
+};
```

Comment:
Can we fold this into `SparseVector` directly? LLVM only does this to save a pointer while making sure that multiple inheritance puts the data array first in the layout of the derived class. I don't think the specialization `SparseVectorStorage<Type, 0>` works for us, `SparseVector` attempts to access `fData` and `fMasks` regardless of the size.

Reply:
Sure, I'll only add an extra static_assert against 0 capacity (that we won't have to support).

Reply:
No, I meant to completely remove `SparseVectorStorage` and put the members into `SparseVector`, or is there a reason this needs to be a separate class? Also `Type fData[N]` should be easier than `char` + `alignas`.

Reply:
OK, I see. I'm getting some errors doing that, but I'll try to fix them.

Reply:
Got rid of SparseVectorStorage but I want to avoid Type fData[N] which calls the default ctor.